src/main/scala/workflow/graph/PipelineEnv.scala

```diff
@@ -30,6 +30,14 @@ class PipelineEnv {
   def setOptimizer(optimizer: Optimizer): Unit = {
     _optimizer = optimizer
   }
+
+  /**
+   * Reset this PipelineEnv (clear state and set the Optimizer to the DefaultOptimizer)
+   */
+  def reset(): Unit = {
+    state.clear()
+    setOptimizer(DefaultOptimizer)
```

Comment:
I had to put this here because state is marked `private[graph]`, so the `PipelineContext` can't access it to reset it.